### PR TITLE
ref(notifications): Document Notification Platform

### PR DESCRIPTION
**Pull request by mgaeta**
The Ecosystem Team is starting to add more documentation to the Notification Platform. In this PR we combine the notions of `type`, `category`, `metrics_key` and `referrer_base`.

Blocked by https://github.com/getsentry/getsentry/pull/7304, ~https://github.com/getsentry/sentry/pull/33381~.

### CODE REVIEW

File: src/sentry/notifications/notifications/organization_request/base.py [hunk 1]
```diff
@@ -65,3 +65,23 @@ def get_log_params(self, recipient: Team | User) -> MutableMapping[str, Any]:
...
+    @property
+    def analytics_event(self) -> str | None:
+        """
+        Override this property with the desired name of an analytics event and
+        it will be recorded in `notify()`. Otherwise, only the generic analytics
+        will be recorded by the providers.
+        """
+        return None
+
+    def record_notification_sent(self, recipient: Team | User, provider: ExternalProviders) -> None:
```
Comment (scefali) [inline src/sentry/notifications/notifications/organization_request/base.py:78]: @mgaeta why are we removing this from the base class?

File: src/sentry/notifications/notifications/organization_request/invite_request.py [hunk 2]
```diff
@@ -13,7 +13,7 @@
 @register()
 class InviteRequestNotification(AbstractInviteRequestNotification):
     analytics_event = "invite_request.sent"
-    referrer_base = "invite_request"
+    metrics_key = "invite-request"
```
Comment (scefali) [inline src/sentry/notifications/notifications/organization_request/invite_request.py:16]: @mgaeta this is going to break a bunch of existing analytics as we are checking for `invite_request` as part of the referrer

**Review by scefali — CHANGES_REQUESTED** [review]

File: src/sentry/notifications/notifications/organization_request/invite_request.py (hunk 2 above)
Comment (scefali) [reply src/sentry/notifications/notifications/organization_request/invite_request.py:16]: This is true for many of the other event types as well

File: src/sentry/notifications/notifications/activity/assigned.py [hunk 3]
```diff
@@ -8,7 +8,7 @@
 
 
 class AssignedActivityNotification(GroupActivityNotification):
-    referrer_base = "assigned-activity"
+    metrics_key = "assigned_activity"
```
Comment (scefali) [inline src/sentry/notifications/notifications/activity/assigned.py:11]: @mgaeta I like the idea of combining everything into one field, but maybe we could pick a different variable name? Like `notification_slug` or something, not sure. But since this field is used for more than metrics, I think the variable name should match.

Comment (ceorourke) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: Maybe `notification_category`?

Comment (scefali) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: Maybe `category`?

File: src/sentry/notifications/notifications/base.py [hunk 4]
```diff
@@ -103,63 +106,43 @@ def get_message_description(self, recipient: Team | User) -> Any:
...
-        Returns an instance for that can be used for analytics such as an organization or project
+        Record an analytics event for this notification. This is called by each
+        provider in `notify()`.
         """
-        return None
-
-    def record_analytics(self, event_name: str, *args: Any, **kwargs: Any) -> None:
-        analytics.record(event_name, *args, **kwargs)
-
-    def record_notification_sent(self, recipient: Team | User, provider: ExternalProviders) -> None:
```
Comment (scefali) [inline src/sentry/notifications/notifications/base.py:141]: @mgaeta Can you pull out the analytics changes into a separate PR? I'd like review that separately.

File: src/sentry/notifications/notifications/activity/assigned.py (hunk 3 above)
Comment (mgaeta) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: I prefer this key to only be used for internal metrics. The description says:
> When we want to collect analytics about this type of notification, we will use this key.

Comment (scefali) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: @mgaeta I'm confused, a change in this PR is making `metrics_key` be used for the referrer query param: https://github.com/getsentry/sentry/pull/33426/files#diff-cc93601c6e7207ef811a9e5b844d97a04dcce9f5e740cb0666680976888e587fR145

I'm fine with a single source of truth for these fields (I think it's a good optimization), just need variable names to make sense

Comment (mgaeta) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: The `referrer` query parameter is being used to collect metrics.

Comment (scefali) [reply src/sentry/notifications/notifications/activity/assigned.py:11]: Sure, fair enough. We can leave it as `metrics_key`

File: src/sentry/notifications/notifications/organization_request/base.py (hunk 1 above)
Comment (scefali) [reply src/sentry/notifications/notifications/organization_request/base.py:78]: @mgaeta can we avoid modifying `record_notification_sent`? It's already a big PR

File: src/sentry/notifications/notifications/base.py [hunk 5]
```diff
@@ -108,63 +111,43 @@ def get_message_description(self, recipient: Team | User) -> Any:
...
-                )
 
     def get_referrer(
         self, provider: ExternalProviders, recipient: Optional[Team | User] = None
     ) -> str:
         # referrer needs the provider and recipient
-        referrer = f"{self.referrer_base}-{EXTERNAL_PROVIDERS[provider]}"
+        referrer = f"{self.metrics_key}-{EXTERNAL_PROVIDERS[provider]}"
         if recipient:
             referrer += "-" + recipient.__class__.__name__.lower()
```
Comment (leeandher) [inline src/sentry/notifications/notifications/base.py:152]: This would also need to be modified, to use `_` instead.

File: tests/sentry/integrations/slack/notifications/test_assigned.py [hunk 6]
```diff
@@ -142,5 +142,5 @@ def test_assignment(self, mock_func):
         assert attachment["title"] == self.group.title
         assert (
             attachment["footer"]
-            == f"{self.project.slug} | <http://testserver/settings/account/notifications/workflow/?referrer=assigned-activity-slack-user|Notification Settings>"
+            == f"{self.project.slug} | <http://testserver/settings/account/notifications/workflow/?referrer=assigned_activity-slack-user|Notification Settings>"
```
[the remark this answers is not in the document]
Comment (scefali) [reply tests/sentry/integrations/slack/notifications/test_assigned.py:145]: @leeandher The referrer is part of the spec because we use it for analytics. If you change the referrer, a lot of charts will break or require changes.

File: src/sentry/notifications/notifications/base.py [hunk 7]
```diff
@@ -108,63 +111,43 @@ def get_message_description(self, recipient: Team | User) -> Any:
...
-                    providers=provider.name.lower(),
-                    **self.get_custom_analytics_params(recipient),
-                )
 
     def get_referrer(
         self, provider: ExternalProviders, recipient: Optional[Team | User] = None
     ) -> str:
         # referrer needs the provider and recipient
-        referrer = f"{self.referrer_base}-{EXTERNAL_PROVIDERS[provider]}"
+        referrer = f"{self.metrics_key}-{EXTERNAL_PROVIDERS[provider]}"
```
[the remark this answers is not in the document]
Comment (scefali) [reply src/sentry/notifications/notifications/base.py:150]: @leeandher It will break a lot of existing analytics if we do this. I agree it's aesthetically unpleasing but there is a major cost to doing it. I'm not saying I'm totally against it but we need to consider the pros vs cons of each.

File: src/sentry/notifications/notifications/organization_request/invite_request.py (hunk 2 above)
Comment (brianthi) [reply src/sentry/notifications/notifications/organization_request/invite_request.py]: If updating the value for metrics_key is going to break analytics, it should not be changed.  If this is not conforming to a stylistic standard we have, then should be addressed separately.  As it is, I don't see why this change is needed for API-2511
